File: 21.kt
import kotlin.math.pow

fun main() {
    val up = Pair(0, -1); val right = Pair(1, 0); val down = Pair(0, 1); val left = Pair(-1, 0)
    val rocks = HashSet<Pair<Int, Int>>()
    var rows = 0

    fun reachable(from: Pair<Int, Int>, steps: Int): Int {
        var reach = HashSet<Pair<Int, Int>>()
        reach.add(from)
        for (i in 1..steps) {
            val next = HashSet<Pair<Int, Int>>()
            fun move(p: Pair<Int, Int>, dir: Pair<Int, Int>) {
                if (!rocks.contains(Pair(p.first + dir.first, p.second + dir.second))) {
                    next.add(Pair(p.first + dir.first, p.second + dir.second))
                }
            }
            for (p in reach) {
                move(p, up); move(p, right); move(p, down); move(p, left)
            }
            reach = next
        }
        return reach.count { it.first >= 0 && it.second >= 0 && it.first < rows && it.second < rows }
    }
    var start = Pair(0, 0)
    while (true) {
        val l = readlnOrNull() ?: break
        l.forEachIndexed { col, ch ->
            when (ch) { 'S' -> start = Pair(col, rows); '#' -> rocks.add(Pair(col, rows)) }
        }
        rows++
    }
    var grids: Long = 26501365.toLong() / rows - 1
    var oddGrids: Long = (grids / 2 * 2 + 1).toDouble().pow(2).toLong()
    var evenGrids: Long = ((grids + 1) / 2 * 2).toDouble().pow(2).toLong()
    var res2: Long = oddGrids * reachable(start, rows * 2 + 1) +
            evenGrids * reachable(start, rows * 2) +
            reachable(Pair(start.first, rows - 1), rows - 1) +
            reachable(Pair(0, start.second), rows - 1) +
            reachable(Pair(start.first, 0), rows - 1) +
            reachable(Pair(rows - 1, start.second), rows - 1) +
            ((grids + 1) * (reachable(Pair(0, rows - 1), rows / 2 - 1) +
                reachable(Pair(rows - 1, rows - 1), rows / 2 - 1) +
                reachable(Pair(0, 0), rows / 2 - 1) +
                reachable(Pair(rows - 1, 0), rows / 2 - 1))) +
            (grids * (reachable(Pair(0, rows - 1), rows * 3 / 2 - 1) +
                reachable(Pair(rows - 1, rows - 1), rows * 3 / 2 - 1) +
                reachable(Pair(0, 0), rows * 3 / 2 - 1) +
                reachable(Pair(rows - 1, 0), rows * 3 / 2 - 1)))
    println(listOf(reachable(start, 64), res2))
}

File: 21.go
package main

import "strings"
import "bufio"
import "fmt"
import "os"

var grid []string

type pos struct {
	row, col int
}

func count(from pos, steps int) int64 {
	var reach = make(map[pos]bool)
	reach[from] = true
	for i := 1; i <= steps; i++ {
		var nexts = make(map[pos]bool)
		for p, _ := range reach {
			for _, dir := range []struct { dr, dc int }{{0, 1}, {0, -1}, {1, 0}, {-1, 0}} {
				var next = pos{p.row + dir.dr, p.col + dir.dc}
				if next.row >= 0 && next.col >= 0 && next.row < len(grid) && next.col < len(grid[0]) &&
					grid[next.row][next.col] != '#' {
					nexts[next] = true
				}
			}
		}
		reach = nexts
	}
	return int64(len(reach))
}

func main() {
	var start pos
	var scanner = bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		if strings.Contains(scanner.Text(), "S") {
			start = pos{len(grid), strings.Index(scanner.Text(), "S")}
		}
		grid = append(grid, scanner.Text())
	}
    var grids = int64(26501365) / int64(len(grid)) - 1
    var oddGrids = (grids / 2 * 2 + 1) * ((grids / 2 * 2 + 1))
    var evenGrids = ((grids + 1) / 2 * 2) * ((grids + 1) / 2 * 2)
	fmt.Println(count(start, 64),
    	oddGrids * count(start, len(grid) * 2 + 1) +
            evenGrids * count(start, len(grid) * 2) +
            count(pos{len(grid) - 1, start.col}, len(grid) - 1) +
            count(pos{start.row, 0}, len(grid) - 1) +
            count(pos{0, start.col}, len(grid) - 1) +
            count(pos{start.row, len(grid) - 1}, len(grid) - 1) +
            ((grids + 1) * (count(pos{0, len(grid) - 1}, len(grid) / 2 - 1) +
                count(pos{len(grid) - 1, len(grid) - 1}, len(grid) / 2 - 1) +
                count(pos{0, 0}, len(grid) / 2 - 1) +
                count(pos{len(grid) - 1, 0}, len(grid) / 2 - 1))) +
            (grids * (count(pos{0, len(grid) - 1}, len(grid) * 3 / 2 - 1) +
                count(pos{len(grid) - 1, len(grid) - 1}, len(grid) * 3 / 2 - 1) +
                count(pos{0, 0}, len(grid) * 3 / 2 - 1) +
                count(pos{len(grid) - 1, 0}, len(grid) * 3 / 2 - 1))),
	)
}
